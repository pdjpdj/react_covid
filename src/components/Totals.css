.totals-table {
  border: 1px solid black;
  width: 20rem;
  margin: auto;
}

.totals-header {
  background-color: darkblue;
  color: white;
  font-weight: bold;
  padding: 1rem 0;
}

.totals-subheader {
  background-color: blue;
  color: white;
  font-weight: bold;
  padding: 0.5rem 0;
}

.table-row {
  display: flex;
  justify-content: center;
  background-color: aliceblue;
  padding: 0.5rem 0;
}

.table-row-alt {
  background-color: white;
}

.table-row-selected {
  background-color: palevioletred;
}

.table-row-label {
  text-align: right;
  padding-right: 0.5rem;
  width: 10rem;
}

.table-row-data {
  padding-left: 0.5rem;
  text-align: left;
  width: 10rem;
}