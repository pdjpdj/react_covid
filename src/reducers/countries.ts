export interface CountryItem {
  Country: string;
  Slug: string;
  ISO2: string;
  latitude: number;
  longitude: number;
}

export interface RapidCountryItem {
  Name: string;
  alpha2code: string;
  alpha3code: string;
  latitude: number;
  longitude: number;
}

interface CountriesState {
  loading: boolean;
  error: boolean;
  countries: CountryItem[] | undefined;
}

export interface CountriesAction {
  type: string;
  payload?: CountryItem[];
}

const countriesActionTypes = {
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  FAILURE: 'FAILURE',
};

export const countriesActionCreators = {
  loading: () => ({type: countriesActionTypes.LOADING}),
  failure: () => ({type: countriesActionTypes.FAILURE}),
  success: (payload: CountryItem[]) => ({
    type: countriesActionTypes.SUCCESS,
    payload,
  }),
};

export const countriesInitialState: CountriesState = {
  loading: true,
  error: false,
  countries: undefined,
};

export function countriesReducer(
  state: CountriesState,
  action: CountriesAction,
): CountriesState {
  switch (action.type) {
    case countriesActionTypes.LOADING:
      return {...state, loading: true, error: false};
    case countriesActionTypes.FAILURE:
      return {...state, loading: false, error: true};
    case countriesActionTypes.SUCCESS:
      return {
        ...state,
        loading: false,
        error: false,
        countries: action.payload,
      };
    default:
      return state;
  }
}
