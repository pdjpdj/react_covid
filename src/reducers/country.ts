export interface CountryTotalItem {
  Country: string;
  CountryCode: string;
  Province: string;
  City: string;
  CityCode: string;
  Confirmed: number;
  Deaths: number;
  Recovered: number;
  Active: number;
  Date: string;
}

interface CountryState {
  loading: boolean;
  error: boolean;
  totals: CountryTotalItem[] | undefined;
}

export interface CountryAction {
  type: string;
  payload?: CountryTotalItem[];
}

const countryActionTypes = {
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  FAILURE: 'FAILURE',
};

export const countryActionCreators = {
  loading: () => ({type: countryActionTypes.LOADING}),
  failure: () => ({type: countryActionTypes.FAILURE}),
  success: (payload: CountryTotalItem[]) => ({
    type: countryActionTypes.SUCCESS,
    payload,
  }),
};

export const countryInitialState: CountryState = {
  loading: true,
  error: false,
  totals: [],
};

export function countryReducer(
  state: CountryState,
  action: CountryAction,
): CountryState {
  switch (action.type) {
    case countryActionTypes.LOADING:
      return {...state, loading: true, error: false};
    case countryActionTypes.FAILURE:
      return {...state, loading: false, error: true};
    case countryActionTypes.SUCCESS:
      return {...state, loading: false, error: false, totals: action.payload};
    default:
      return state;
  }
}
