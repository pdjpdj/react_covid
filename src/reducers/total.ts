export interface TotalItem {
  NewConfirmed: number;
  TotalConfirmed: number;
  NewDeaths: number;
  TotalDeaths: number;
  NewRecovered: number;
  TotalRecovered: number;
}

interface TotalState {
  loading: boolean;
  error: boolean;
  totals: TotalItem | undefined;
}

export interface TotalAction {
  type: string;
  payload?: TotalItem;
}

const totalActionTypes = {
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  FAILURE: 'FAILURE',
};

export const totalActionCreators = {
  loading: () => ({type: totalActionTypes.LOADING}),
  failure: () => ({type: totalActionTypes.FAILURE}),
  success: (payload: TotalItem) => ({type: totalActionTypes.SUCCESS, payload}),
};

export const totalInitialState: TotalState = {
  loading: true,
  error: false,
  totals: undefined,
};

export function totalReducer(
  state: TotalState,
  action: TotalAction,
): TotalState {
  switch (action.type) {
    case totalActionTypes.LOADING:
      return {...state, loading: true, error: false};
    case totalActionTypes.FAILURE:
      return {...state, loading: false, error: true};
    case totalActionTypes.SUCCESS:
      return {...state, loading: false, error: false, totals: action.payload};
    default:
      return state;
  }
}
