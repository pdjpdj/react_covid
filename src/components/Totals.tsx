import React from 'react';
import {TotalItem} from '../reducers/total';

interface TotalsProps {
  totals: TotalItem | undefined;
}

function Totals({totals}: TotalsProps) {
  if (!totals) {
    return (
      <div>
        <div>
          No global data
        </div>
      </div>
    );
  }

  return (
    <div>
      <table>
        <caption className='totals-subheader'>
          Today
        </caption>
        <tbody>
          <tr className='table-row'>
            <td className='table-row-label'>
              Confirmed:
            </td>
            <td className='table-row-data'>
              {totals?.NewConfirmed.toLocaleString()}
            </td>
          </tr>
          <tr className='table-row table-row-alt'>
            <td className='table-row-label'>
              Deaths:
            </td>
            <td className='table-row-data'>
              {totals.NewDeaths.toLocaleString()}
            </td>
          </tr>
          <tr className='table-row'>
            <td className='table-row-label'>
              Recovered:
            </td>
            <td className='table-row-data'>
              {totals?.NewRecovered.toLocaleString()}
            </td>
          </tr>
        </tbody>
      </table>
      <table>
        <caption className='totals-subheader'>
          All time
        </caption>
        <tbody>
          <tr className='table-row'>
            <td className='table-row-label'>
              Confirmed:
            </td>
            <td className='table-row-data'>
              {totals?.TotalConfirmed.toLocaleString()}
            </td>
          </tr>
          <tr className='table-row table-row-alt'>
            <td className='table-row-label'>
              Deaths:
            </td>
            <td className='table-row-data'>
              {totals.TotalDeaths.toLocaleString()}
            </td>
          </tr>
          <tr className='table-row'>
            <td className='table-row-label'>
              Recovered:
            </td>
            <td className='table-row-data'>
              {totals?.TotalRecovered.toLocaleString()}
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  );
}

export default Totals;
