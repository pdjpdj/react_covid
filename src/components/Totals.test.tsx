import React from 'react';
import { render, screen } from '@testing-library/react';
import Totals from './Totals';
import { TotalItem } from '../reducers/total';

const mockTotal: TotalItem = 
  {
    NewConfirmed: 4500,
    NewDeaths: 123,
    NewRecovered: 340,
    TotalConfirmed: 4500,
    TotalDeaths: 123,
    TotalRecovered: 340,
  };

test('renders country totals with data matches snapshot', () => {
  const {container} = render(<Totals totals={mockTotal} />);
  expect(container.firstChild).toMatchSnapshot();
});

test('renders no data when no table undefined', () => {
  render(<Totals totals={undefined} />);
  const text = screen.getAllByText(/^No global data$/);
  expect(text.length).toEqual(1);
});