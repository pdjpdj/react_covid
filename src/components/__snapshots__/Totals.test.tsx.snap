// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders country totals with data matches snapshot 1`] = `
<div>
  <table>
    <caption
      class="totals-subheader"
    >
      Today
    </caption>
    <tbody>
      <tr
        class="table-row"
      >
        <td
          class="table-row-label"
        >
          Confirmed:
        </td>
        <td
          class="table-row-data"
        >
          4,500
        </td>
      </tr>
      <tr
        class="table-row table-row-alt"
      >
        <td
          class="table-row-label"
        >
          Deaths:
        </td>
        <td
          class="table-row-data"
        >
          123
        </td>
      </tr>
      <tr
        class="table-row"
      >
        <td
          class="table-row-label"
        >
          Recovered:
        </td>
        <td
          class="table-row-data"
        >
          340
        </td>
      </tr>
    </tbody>
  </table>
  <table>
    <caption
      class="totals-subheader"
    >
      All time
    </caption>
    <tbody>
      <tr
        class="table-row"
      >
        <td
          class="table-row-label"
        >
          Confirmed:
        </td>
        <td
          class="table-row-data"
        >
          4,500
        </td>
      </tr>
      <tr
        class="table-row table-row-alt"
      >
        <td
          class="table-row-label"
        >
          Deaths:
        </td>
        <td
          class="table-row-data"
        >
          123
        </td>
      </tr>
      <tr
        class="table-row"
      >
        <td
          class="table-row-label"
        >
          Recovered:
        </td>
        <td
          class="table-row-data"
        >
          340
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
